common/determined_common/experimental/experiment.py

```diff
@@ -68,17 +69,33 @@ def top_n_checkpoints(
                 this parameter is ignored. By default, the value of ``smaller_is_better``
                 from the experiment's configuration is used.
         """
-        r = api.get(self._master, "checkpoints", params={"experiment_id": self.id}).json()
+        r = api.get(
+            self._master,
+            "/api/v1/experiments/{}/checkpoints".format(self.id),
+            params={
+                "states": checkpoint.CheckpointState.COMPLETED.value,
+                "validation_states": checkpoint.CheckpointState.COMPLETED.value,
+            },
+        )
+        checkpoints = r.json()["checkpoints"]
 
         if not r:
             raise AssertionError("No checkpoint found for trial {}".format(self.id))
 
         if not sort_by:
-            sort_by = r[0]["experiment_config"]["searcher"]["metric"]
-            smaller_is_better = r[0]["experiment_config"]["searcher"]["smaller_is_better"]
+            sort_by = checkpoints[0]["experimentConfig"]["searcher"]["metric"]
+            smaller_is_better = checkpoints[0]["experimentConfig"]["searcher"]["smaller_is_better"]
 
-        r.sort(
-            reverse=not smaller_is_better, key=lambda x: x["metrics"]["validation_metrics"][sort_by]
+        checkpoints.sort(
+            reverse=not smaller_is_better, key=lambda x: x["metrics"]["validationMetrics"][sort_by]
         )
 
-        return [checkpoint.Checkpoint.from_json(ckpt, self._master) for ckpt in r[:limit]]
+        # Ensure returned checkpoints are from distinct trials.
+        t_ids = set()
+        checkpoint_refs = []
+        for ckpt in checkpoints[:limit]:
```

Comment:
No this is not the case. We do not send the limit param to the endpoint so we should receive all checkpoints for the experiment that are completed and have validation steps associated. Here we are taking the "best" checkpoint per trial and throwing away the other checkpoints from the same trial. Something to note is that this method was supposed to return the best n checkpoints for distinct trials but there was a regression when we moved off of graphql.

Reply:
Talked off line, since we sort by `validationMetrics` then take the first `limit`, if one trial had really good metrics, it could be the first `limit` checkpoints.